src/node/stage/approved.rs

```diff
@@ -1895,318 +1806,62 @@ impl ApprovedPeer {
-                    | Variant::ParsecResponse(..)
-                    | Variant::Ping
-                    | Variant::Bounce { .. } => false,
-                }
-            }
-
-            Variant::BootstrapResponse(_) | Variant::NodeApproval(_) | Variant::Ping => false,
+    // TODO: move to Chain
+    fn print_rt_size(&self) {
+        const TABLE_LVL: log::Level = log::Level::Info;
+        if log::log_enabled!(TABLE_LVL) {
+            let status_str = format!("Routing Table size: {:3}", self.chain.elders().count());
+            let network_estimate = match self.chain.network_size_estimate() {
+                (n, true) => format!("Exact network size: {}", n),
+                (n, false) => format!("Estimated network size: {}", n),
+            };
+            let sep_len = cmp::max(status_str.len(), network_estimate.len());
+            let sep_str = iter::repeat('-').take(sep_len).collect::<String>();
+            log!(target: "routing_stats", TABLE_LVL, " -{}- ", sep_str);
+            log!(target: "routing_stats", TABLE_LVL, "| {:<1$} |", status_str, sep_len);
+            log!(target: "routing_stats", TABLE_LVL, "| {:<1$} |", network_estimate, sep_len);
+            log!(target: "routing_stats", TABLE_LVL, " -{}- ", sep_str);
         }
     }
-
-    fn verify_message(&self, msg: &Message) -> Result<bool, RoutingError> {
-        self.verify_message_quiet(msg).map_err(|error| {
-            self.log_verify_failure(msg, &error, self.chain.get_their_key_infos());
-            error
-        })
-    }
 }
 
-#[cfg(feature = "mock_base")]
-impl ApprovedPeer {
-    pub fn chain(&self) -> &Chain {
-        &self.chain
-    }
-
-    pub fn has_unpolled_observations(&self) -> bool {
-        self.parsec_map.has_unpolled_observations()
-    }
-
-    pub fn unpolled_observations_string(&self) -> String {
-        self.parsec_map.unpolled_observations_string()
-    }
-
-    pub fn is_peer_our_elder(&self, pub_id: &PublicId) -> bool {
-        self.chain.is_peer_our_elder(pub_id)
-    }
+#[derive(Debug, PartialEq, Eq, PartialOrd, Ord, Hash, Clone)]
+enum PendingMessageKey {
+    NeighbourInfo {
+        version: u64,
+        prefix: Prefix<XorName>,
+    },
+}
 
-    pub fn send_msg_to_targets(
-        &mut self,
-        dst_targets: &[SocketAddr],
-        dg_size: usize,
-        message: Message,
-    ) -> Result<(), RoutingError> {
-        let message = message.to_bytes()?;
-        self.send_message_to_targets(dst_targets, dg_size, message);
-        Ok(())
-    }
+struct CompleteParsecReset {
+    // The new genesis prefix info.
+    gen_pfx_info: GenesisPfxInfo,
+    // The cached events that should be revoted. Not shared state: only the ones we voted for.
+    // Also contains our votes that never reached consensus.
+    to_vote_again: BTreeSet<NetworkEvent>,
+    // The events to process. Not shared state: only the ones we voted for.
+    // Also contains our votes that never reached consensus.
```

Comment:
so both `to_vote_again` and `to_process` contains `our votes that never reached consensus` ?

Reply:
Yes. Both these collections contains events from the previous parsec instance that have not reached consensus yet. Some of those events might be in `to_process`, some in `to_vote_again`, some in both and some in neither.